src/server/pkg/work/work.go

```diff
@@ -14,66 +17,117 @@ import (
-// them upon completion. The collectFunc callback will be called for
-// each subtask that is collected.
-type Master struct {
-	etcdClient          *etcd.Client
-	taskCol, subtaskCol col.Collection
-	collectFunc         CollectFunc
-}
-
-// NewMaster creates a new master.
-func NewMaster(etcdClient *etcd.Client, etcdPrefix string, collectFunc CollectFunc) *Master {
-	return &Master{
-		etcdClient:  etcdClient,
-		taskCol:     collection(etcdClient, path.Join(etcdPrefix, taskPrefix)),
-		subtaskCol:  collection(etcdClient, path.Join(etcdPrefix, subtaskPrefix)),
-		collectFunc: collectFunc,
+type taskEtcd struct {
+	etcdClient                    *etcd.Client
+	taskCol, subtaskCol, claimCol col.Collection
+}
+
+func NewTaskQueue(ctx context.Context, etcdClient *etcd.Client, etcdPrefix string, taskNamespace string) *TaskQueue {
+	return &TaskQueue{
+		taskEtcd:  newTaskEtcd(etcdClient, etcdPrefix, taskNamespace),
+		taskQueue: newTaskQueue(ctx),
 	}
 }
 
-func collection(etcdClient *etcd.Client, etcdPrefix string) col.Collection {
+func newTaskEtcd(etcdClient *etcd.Client, etcdPrefix string, taskNamespace string) *taskEtcd {
+	return &taskEtcd{
+		etcdClient: etcdClient,
+		taskCol:    newCollection(etcdClient, path.Join(etcdPrefix, taskPrefix, taskNamespace), &Task{}),
+		subtaskCol: newCollection(etcdClient, path.Join(etcdPrefix, subtaskPrefix, taskNamespace), &TaskInfo{}),
+		claimCol:   newCollection(etcdClient, path.Join(etcdPrefix, claimPrefix, taskNamespace), &Claim{}),
+	}
+}
+
+func newCollection(etcdClient *etcd.Client, etcdPrefix string, template proto.Message) col.Collection {
 	return col.NewCollection(
 		etcdClient,
 		etcdPrefix,
 		nil,
-		&TaskInfo{},
+		template,
 		nil,
 		nil,
 	)
 }
 
-// Run runs the master with a given context and task.
-func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
-	taskInfo := &TaskInfo{Task: task}
-	if err := m.updateTaskInfo(ctx, taskInfo); err != nil {
+// RunTask runs a task in the task queue.
+// The task code should be contained within the passed in callback.
+// The callback will receive a Master, which should be used for running subtasks in the task queue.
+// The task state will be cleaned up upon return of the callback.
+func (tq *TaskQueue) RunTask(ctx context.Context, task *Task, f func(*Master)) (retErr error) {
+	if _, err := col.NewSTM(ctx, tq.etcdClient, func(stm col.STM) error {
+		return tq.taskCol.ReadWrite(stm).Put(task.ID, task)
+	}); err != nil {
 		return err
 	}
 	defer func() {
-		taskInfo.State = State_SUCCESS
-		// (bryce) might also want a killed state.
 		if retErr != nil {
-			taskInfo.State = State_FAILURE
+			if err := tq.deleteTask(task.ID); err != nil {
+				fmt.Printf("errored deleting task %v: %v\n", task.ID, err)
+			}
+		}
+	}()
+	return tq.taskQueue.runTask(ctx, task.ID, func(te *taskEntry) {
+		defer func() {
+			if err := tq.deleteTask(task.ID); err != nil {
+				fmt.Printf("errored deleting task %v: %v\n", task.ID, err)
+			}
+		}()
```

Comment:
It looks like `taskQueue.runTask` already cleans up the task before returning - I _think_ both the defers in this function can go away.

Reply:
I think all three deletes are necessary. The one in `runTask` removes the entry in the `taskQueue` (internal data structure that actually implements the queuing behavior). The two in `RunTask` are necessary for cleaning up the etcd state associated with `RunTask`. There is one that handles the case where we error when attempting to setup the task in the internal task queue, but after setting up the etcd entry (The etcd entry for the task needs to be correctly setup before we can run the callback passed in by the user). The other is the happy path for cleanup after the client callback is completed.

Reply:
Ah, I see - I was confused between `taskQueue.deleteTask` and `TaskQueue.deleteTask` - I really think one of these should be renamed...